File: HexChat/twitch.py
import hexchat

__module_name__ = 'Twitch'
__module_author__ = 'TingPing'
__module_version__ = '3'
__module_description__ = 'Better integration with Twitch.tv'
# Very much a work in progress...

# Commands from http://help.twitch.tv/customer/portal/articles/659095-chat-moderation-commands
# /ban may conflict with other scripts nothing we can do about that
# /clear is an existing command, just override it
commands = ('timeout', 'slow', 'slowoff', 'subscribers', 'subscribersoff',
			'mod', 'unmod', 'mods', 'clear', 'ban', 'unban', 'commercial',
			'r9kbeta', 'r9kbetaoff', 'color')

aliases = {'op':'mod', 'deop':'unmod'}

def twitchOnly(func):
	def if_twitch(*args, **kwargs):
		for host in (hexchat.get_info('host'), hexchat.get_info('server')):
			if host and 'twitch.tv' in host:
				return func(*args, **kwargs)
		else:
			return hexchat.EAT_NONE
	return if_twitch

# Twitch returns a lot of 'unknown command' errors, ignore them.
@twitchOnly
def servererr_cb(word, word_eol, userdata):
	if word[3] in ('WHO', 'WHOIS'):
		return hexchat.EAT_ALL

#@twitchOnly
#def endofmotd_cb(word, word_eol, userdata):
	# TODO: This shows the output of commands like /slow but also
	# tons of other garbage we have to ignore..
	# hexchat.command('QUOTE TWITCHCLIENT 3')

# Print jtv messages in server tab.
@twitchOnly
def privmsg_cb(word, word_eol, userdata):
	if word[0][1:].split('!')[0] == 'jtv':
		if word[2][0] != '#':
			for chan in hexchat.get_list('channels'):
				if chan.type == 1 and chan.id == hexchat.get_prefs('id'):
					chan.context.emit_print('Server Text', word_eol[3][1:])
			return hexchat.EAT_ALL
		elif word[3] in (':USERCOLOR', ':USERSTATE', ':HISTORYEND', ':EMOTESET', ':CLEARCHAT', ':SPECIALUSER'):
			return hexchat.EAT_ALL

# Eat any message starting with a '.', twitch eats all of them too.
@twitchOnly
def yourmsg_cb(word, word_eol, userdata):
	if word[1][0] == '.':
		return hexchat.EAT_ALL

# Just prefix with a '.'.
@twitchOnly
def command_cb(word, word_eol, alias):
	if alias:
		if len(word_eol) > 1:
			hexchat.command('say .{} {}'.format(alias, word_eol[1]))
		else:
			hexchat.command('say .{}'.format(alias))
	else:
		hexchat.command('say .{}'.format(word_eol[0]))
	return hexchat.EAT_ALL

for command in commands:
	hexchat.hook_command(command, command_cb)
for command, alias in aliases.items():
	hexchat.hook_command(command, command_cb, alias)
hexchat.hook_print('Your Message', yourmsg_cb)
hexchat.hook_server('421', servererr_cb)
#hexchat.hook_server('376', endofmotd_cb)
hexchat.hook_server('PRIVMSG', privmsg_cb)

File: HexChat/twitch.lua
-- SPDX-License-Identifier: MIT
hexchat.register('Twitch', '1', 'Better integration with twitch.tv')

local function is_twitch ()
	local server = hexchat.get_info('server')
	if hexchat.nickcmp(hexchat.get_info('network'), 'Twitch') == 0 then
		return true
	elseif server and server:find('twitch.tv$') then
		return true
	else
		return false
	end
end


-- Commands from http://help.twitch.tv/customer/portal/articles/659095-chat-moderation-commands
-- /ban may conflict with other scripts nothing we can do about that
-- /clear is an existing command, just override it
for _, command in pairs({
	'timeout', 'slow', 'slowoff', 'subscribers', 'subscribersoff',
	'mod', 'unmod', 'mods', 'clear', 'ban', 'unban', 'commercial',
	'r9kbeta', 'r9kbetaoff', 'color', 'host',
}) do
	hexchat.hook_command(command, function (word, word_eol)
		if not is_twitch() then return end

		hexchat.command('say .' .. word_eol[1])
		return hexchat.EAT_ALL
	end)
end

for command, alias in pairs({['op'] = 'mod', ['deop'] = 'unmod'}) do
	hexchat.hook_command(command, function (word, word_eol)
		if not is_twitch() then return end

		if word[2] then
			hexchat.command(string.format('say .%s %s', alias, word_eol[2]))
		else
			hexchat.command('say .' .. alias)
		end
		return hexchat.EAT_ALL
	end)
end

hexchat.hook_server('PRIVMSG', function (word, word_eol)
	if not is_twitch() then return end

	-- Move jtv messages to the server tab
	if word[1]:find('^:jtv!') and word[3]:sub(1, 1) ~= '#' then
		local id = hexchat.prefs['id']
		for chan in hexchat.iterate('channels') do
			if chan.type == 1 and chan.id == id then
				chan.context:emit_print('Server Text', word_eol[4]:sub(2))
				return hexchat.EAT_ALL
			end
		end
	end
end)

hexchat.hook_server('421', function (word, word_eol)
	if not is_twitch() then return end

	-- Ignore unknown command errors
	if word[4] == 'WHO' or word[4] == 'WHOIS' then
		return hexchat.EAT_ALL
	end
end)

hexchat.hook_print('Your Message', function (args)
	if not is_twitch() then return end

	-- Eat any message starting with a '.', twitch eats all of them too
	if args[2]:sub(1, 1) == '.' then
		return hexchat.EAT_ALL
	end
end)

hexchat.hook_print('Channel Message', function (args)
	if not is_twitch() then return end

	-- Format messages from twitchnotify
	if args[1] == 'twitchnotify' then
		print('\00314*\t' .. args[2])
		return hexchat.EAT_HEXCHAT
	end
end, hexchat.PRI_LOW)
